internal/pkg/service/stream/storage/level/local/diskwriter/network/router/pipeline_balanced.go

```diff
@@ -0,0 +1,71 @@
+package router
+
+import (
+	"context"
+	"sync"
+
+	"golang.org/x/exp/maps"
+
+	"github.com/keboola/keboola-as-code/internal/pkg/service/stream/definition/key"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/stream/mapping/recordctx"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/stream/sink/pipeline"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/stream/storage/level/local/diskwriter/network/router/assignment"
+)
+
+type balancedPipeline struct {
+	router    *Router
+	lock      sync.RWMutex
+	pipelines []SlicePipeline
+}
+
+func newBalancedPipeline(ctx context.Context, router *Router, sinkKey key.SinkKey) (pipeline.Pipeline, error) {
+	sinkSlices, err := router.sinkOpenedSlices(sinkKey)
+	if err != nil {
+		return nil, err
+	}
+
+	assigned := assignment.AssignSlices(
+		maps.Keys(sinkSlices),
+		router.distribution.Nodes(),
+		router.distribution.NodeID(),
+		router.config.MinSlicesPerSourceNode,
+	)
+
+	balanced := &balancedPipeline{}
+	for _, sliceKey := range assigned {
+		// sub.startOpenLoop()
+		balanced.pipelines = append(balanced.pipelines, newSlicePipeline(ctx, router, sinkSlices[sliceKey]))
+	}
+
+	// return balanced, nil
+	return balanced, nil
+}
+
+func (p *balancedPipeline) ReopenOnSinkModification() bool {
+	// We are watching for slices changes, we don't need other reopen mechanism.
+	return false
+}
+
+func (p *balancedPipeline) WriteRecord(c recordctx.Context) (pipeline.RecordStatus, error) {
+	p.lock.RLock()
```

Comment:
https://www.reddit.com/r/golang/comments/15pe2wx/am_i_understanding_the_behavior_of_syncrwmutex/

```
RWMutex is ... write-biased reader-writer lock, designed to avoid writer starvation in case of high reader load.
```